File: aoc/2016/aoc201605.go
package main

import (
	"crypto/md5"
	"fmt"
	"os"
)

func generate(doorID string) string {
	var result [8]byte
	index := 0
	for i := 0; ; i++ {
		hash := md5.Sum([]byte(fmt.Sprintf("%s%d", doorID, i)))
		if hash[0] == 0 && hash[1] == 0 && hash[2] < 16 {
			result[index] = hash[2]
			index++
			if index >= 8 {
				return fmt.Sprintf("%x%x%x%x%x%x%x%x", result[0], result[1], result[2], result[3], result[4], result[5], result[6], result[7])
			}
		}
	}
}

func generate2(doorID string) string {
	var result [8]byte
	var filled [8]bool
iterating:
	for i := 0; ; i++ {
		hash := md5.Sum([]byte(fmt.Sprintf("%s%d", doorID, i)))
		if hash[0] == 0 && hash[1] == 0 && hash[2] < 8 {
			index := int(hash[2])
			if filled[index] {
				continue
			}
			filled[index] = true
			result[index] = hash[3] >> 4
			for _, f := range filled {
				if !f {
					continue iterating
				}
			}
			return fmt.Sprintf("%x%x%x%x%x%x%x%x", result[0], result[1], result[2], result[3], result[4], result[5], result[6], result[7])
		}
	}
	
}

func main() {
	if len(os.Args) < 2 {
		if p := generate("abc"); p != "18f47a30" {
			panic(p)
		}
		if p := generate2("abc"); p != "05ace8e3" {
			panic(p)
		}
	} else if len(os.Args) < 3 {
		println(generate(os.Args[1]))
	} else {
		println(generate2(os.Args[2]))
	}
}

File: aoc/2024/aoc201605.hs
module AOC201605 where

import Crypto.Hash.MD5(hash)
import Data.ByteString(ByteString,index,pack)
import Data.Char(ord)
import Data.Map(Map,elems,empty,insert,member,size)
import Data.Word(Word8)

import AOC

aoc = AOC {
    day="../../2016/input/05",
    aocTests=[
        AOCTest {
            testData=unlines [
                "abc"
                ],
            testResult=Just $ show "18f47a30",
            testResult2=Just $ show "05ace8e3"
            }
        ],
    aocCode=ParallelCode {
        pcodeParse=const parse,
        pcodeParse2=const parse,
        pcodeTest=result,
        pcodeTest2=result2,
        pcodeResult=result,
        pcodeResult2=result2
        }
    }

parse :: String -> [Word8]
parse = map (fromIntegral . ord) . filter (/= '\n')

hex :: Word8 -> Char
hex 0 = '0'
hex 1 = '1'
hex 2 = '2'
hex 3 = '3'
hex 4 = '4'
hex 5 = '5'
hex 6 = '6'
hex 7 = '7'
hex 8 = '8'
hex 9 = '9'
hex 10 = 'a'
hex 11 = 'b'
hex 12 = 'c'
hex 13 = 'd'
hex 14 = 'e'
hex 15 = 'f'

hashes :: Int -> Word8 -> Int -> [Word8] -> [ByteString]
hashes ncpu max6th n doorID =
    parallelMapReduce ncpu generate concat [n..n+1000*ncpu-1]
        ++ hashes ncpu max6th (n+1000*ncpu-1) doorID
  where
    generate i
      | index h 0 == 0 && index h 1 == 0 && index h 2 < max6th = [h]
      | otherwise = []
      where h = hash $ pack $ doorID ++ map (fromIntegral . ord) (show i)

result ncpu = map (hex . (`index` 2)) . take 8 . hashes ncpu 16 0

collect :: Map Word8 Char -> [ByteString] -> String
collect pwd (h:hs)
  | member (index h 2) pwd = collect pwd hs
  | size pwd < 7 = collect nextPwd hs
  | otherwise = elems nextPwd
  where nextPwd = insert (index h 2) (hex (index h 3 `div` 16)) pwd

result2 ncpu = collect empty . hashes ncpu 8 0
